pgxn/neon/libpagestore.c

```diff
@@ -63,81 +54,192 @@ int			flush_every_n_requests = 8;
 }
 
 static void
-ReloadConnstring()
+psm_init(void)
 {
-    if(!pagestore_shared)
-        return;
-    LWLockAcquire(pagestore_shared->lock, LW_SHARED);
-    strlcpy(local_pageserver_connstring, pagestore_shared->pageserver_connstring, sizeof(local_pageserver_connstring));
-    pagestore_local_counter = pg_atomic_read_u64(&pagestore_shared->update_counter);
-    LWLockRelease(pagestore_shared->lock);
+	prev_shmem_startup_hook = shmem_startup_hook;
+	shmem_startup_hook = psm_shmem_startup;
+#if PG_VERSION_NUM>=150000
+	prev_shmem_request_hook = shmem_request_hook;
+	shmem_request_hook = psm_shmem_request;
+#else
+	psm_shmem_request();
+#endif
+}
+
+/*
+ * Reload page map if needed and return number of shards and connection string for the specified shard
+ */
+static shardno_t
+load_shard_map(shardno_t shard_no, char* connstr)
+{
+	shardno_t n_shards;
+
+	LWLockAcquire(shard_map_lock, LW_SHARED);
+
+	while (shard_map->n_shards == 0 || shard_map_update_counter != shard_map->update_counter)
+	{
+		/* Close all existed connections */
+		for (shardno_t i = 0; i < max_attached_shard_no; i++)
+		{
+			if (page_servers[i].conn)
+				pageserver_disconnect(i);
+		}
+
+		/* Request new shard map from control plane under exclusive lock */
+		LWLockRelease(shard_map_lock);
+		LWLockAcquire(shard_map_lock, LW_EXCLUSIVE);
+		if (shard_map->n_shards == 0)
+		{
+			char const* shard_connstr = page_server_connstring;
+			char const* sep;
+			size_t connstr_len;
+			do
+			{
+				sep = strchr(shard_connstr, ',');
+				connstr_len = sep != NULL ? sep - shard_connstr : strlen(shard_connstr);
+				if (connstr_len == 0)
+					break; /* trailing comma */
+				if (shard_map->n_shards >= MAX_SHARDS)
+					elog(ERROR, "Too many shards");
+				if (connstr_len >= MAX_PS_CONNSTR_LEN)
+					elog(ERROR, "Connection  string too long");
+				memcpy(shard_map->shard_connstr[shard_map->n_shards++], shard_connstr, connstr_len+1);
+				shard_connstr = sep + 1;
+			} while (sep != NULL);
+
+			if (shard_map->n_shards == 0)
+				elog(ERROR, "No shards were speciified");
+
+			shard_map->update_counter += 1;
+		}
+		shard_map_update_counter = shard_map->update_counter;
+	}
+	n_shards = shard_map->n_shards;
+	if (shard_no >= n_shards)
+		elog(ERROR, "Shard %d is greater or equal than number of shards %d", shard_no, n_shards);
+
+	if (connstr)
+		strcpy(connstr, shard_map->shard_connstr[shard_no]);
+
+	LWLockRelease(shard_map_lock);
+
+	return n_shards;
+}
+
+#define MB (1024*1024)
+
+shardno_t
+get_shard_number(BufferTag* tag)
+{
+	shardno_t n_shards = load_shard_map(0, NULL);
+	uint32	  hash;
+
+#if PG_MAJORVERSION_NUM < 16
+	hash = murmurhash32(tag->rnode.spcNode);
+	hash_combine(hash, murmurhash32(tag->rnode.dbNode));
+	hash_combine(hash, murmurhash32(tag->rnode.relNode));
+	hash_combine(hash, murmurhash32(tag->blockNum/(MB/BLCKSZ)/stripe_size));
+#else
+	hash = murmurhash32(tag->spcOid);
+	hash_combine(hash, murmurhash32(tag->dbOid));
```

Comment:
These hash_combine lines need a `hash = ` at the start

Reply:
Sorry, I do not understand it. There is  `hash = murmurhash32(tag->spcOid);` at start

Reply:
`hash_combine` takes its arguments by value and returns the combined result -- so to achieve the desired result it should be `hash =  hash_combine(hash, ...` etc.

Reply:
Brrr...
This is why reviews are needed:)
I so many times looked at this code that just don't notice this obvious bug (may be because of `hash_` prefix:)
Thank you, fixed.